backend/ranking.js

```diff
@@ -3,10 +3,184 @@ import { Follow } from "./models/follow.js";
+                const stocksTheyLike = friendLikes.map((friendLike) => {
+                    return friendLike.Stock.dataValues;
+                });
+                return stocksTheyLike;
+            })
+        );
+
+        const stocksThatFriendsDislike = await Promise.all(
+            friendIDs.map(async (friendID) => {
+                const friendDislikes = await Dislike.findAll({
+                    where: { UserId: friendID },
+                    include: [{ model: Stock }]
+                });
+                const stocksTheyDislike = friendDislikes.map((friendDislike) => {
+                    return friendDislike.Stock.dataValues;
+                });
+                return stocksTheyDislike;
+            })
+        );
+
+        const friendsFollowsMap = {};
+
+        stocksThatFriendsFollow.forEach((friendStocks) => {
+            friendStocks.forEach((stock) => {
+                if (friendsFollowsMap[stock.ticker]) {
+                    friendsFollowsMap[stock.ticker] += 1;
+                } else {
+                    friendsFollowsMap[stock.ticker] = 1;
+                }
+            });
+        });
+
+        const friendsLikesMap = {};
+
+        stocksThatFriendsLike.forEach((friendStocks) => {
+            friendStocks.forEach((stock) => {
+                if (friendsLikesMap[stock.ticker]) {
+                    friendsLikesMap[stock.ticker] += 1;
+                } else {
+                    friendsLikesMap[stock.ticker] = 1;
+                }
+            });
+        });
+
+        const friendsDislikesMap = {};
+
+        stocksThatFriendsDislike.forEach((friendStocks) => {
+            friendStocks.forEach((stock) => {
+                if (friendsDislikesMap[stock.ticker]) {
+                    friendsDislikesMap[stock.ticker] += 1;
+                } else {
+                    friendsDislikesMap[stock.ticker] = 1;
+                }
+            });
+        });
+
+        const stocksRanking = stocksInDatabase.map((currentStock) => {
+            let points = 0;
+
+            if (followsMap[currentStock.ticker]) {
+                points += FOLLOW_POINTS;
+            }
+
+            if (likesMap[currentStock.ticker]) {
+                points += LIKE_POINTS;
+            }
+
+            if (dislikesMap[currentStock.ticker]) {
+                points += DISLIKE_POINTS;
+            }
+
+            if (friendsFollowsMap[currentStock.ticker]) {
+                points += friendsFollowsMap[currentStock.ticker] * FOLLOW_FRIEND_POINTS;
+            }
+
+            if (friendsLikesMap[currentStock.ticker]) {
+                points += friendsLikesMap[currentStock.ticker] * LIKE_FRIEND_POINTS;
+            }
+
+            if (friendsDislikesMap[currentStock.ticker]) {
+                points +=
+                    friendsDislikesMap[currentStock.ticker] *
+                    DISLIKE_FRIEND_POINTS *
+                    (friendsDislikesMap[currentStock.ticker] % 2 === 0 ? -1 : 1);
```

Comment:
Can you explain why using mod 2 to calculate the dislikes points?

Reply:
DISLIKE_FRIEND_POINTS is a negative number and so if you multiply a negative by an even number it becomes positive, so it would be like adding points to that stock even though we should be subtracting them.